### ISSUE TITLE
Add Type Hints to operations, facts & connectors

### ISSUE THREAD

**Issue by stone-w4tch3r** [Fizzadar/pyinfra#1074]
What about adding type hints to API? For example, operations will benefit greatly.

for example, apt:
```py
@operation
def packages(
    packages: list[str] = None,
    present=True,
    latest=False,
    update=False,
    cache_time: int = None,
    upgrade=False,
    force=False,
    no_recommends=False,
    allow_downgrades=False,
    extra_install_args: list[str] = None,
    extra_uninstall_args: list[str] = None,
):
    pass
```

**Comment by Fizzadar** [comment]
Absolutely! Now that the core api is (mostly) typed it’s time to start working through operations. Nice thing is each can be done in isolation, welcoming PRs :)

**Comment by jinnatar** [comment]
Related to being able to type operations in the first place, is the `@operation()` decorator itself sufficiently typed to actually enable that? I'm having a real hard time typing a custom operation because the decorator being "partially unknown" taints what it wraps, meaning type hinting for my operation is completely ignored when writing downstream tasks that use it.

> basedpyright: Type of "operation" is partially unknown                                                                                                                            
Type of "operation" is "(is_idempotent: bool = True, idempotent_notice: str | None = None, is_deprecated: bool = False, deprecated_for: str | None = None, _set_in_op: bool = True)
-> ((((**P@operation) -> Generator[Unknown, None, None])) -> PyinfraOperation[P@operation])" [reportUnknownVariableType]
